File: 哈希表/73. 矩阵置零/矩阵置零.cpp
#include<iostream>
#include<vector>
using namespace std;
// 上面的头文件与题解无光,只是为了在本地编译器里不报错

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int weightlength = matrix[0].size();
        int heightlength = matrix.size();
        vector<bool>weiight(weightlength,false);
        vector<bool>height(heightlength,false);
        for(int i=0;i<heightlength;i++){
            for(int j=0;j<weightlength;j++){
                if(matrix[i][j]==0){
                    weiight[j]=true;
                    height[i]=true;
                }
            }
        }
        for(int i=0;i<heightlength;i++){
            for(int j=0;j<weightlength;j++){
                if(weiight[j]||height[i]){
                    matrix[i][j]=0;
                }
            }
        }
    }
};

File: 哈希表/73. 矩阵置零/矩阵置零.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var setZeroes = function (matrix) {
  // 使用一维数组的写法就可以写出来这个题目
  // 但是这个题目要求使用常数空间，所以我们需要使用一个额外的标记来记录第一行和第一列是否需要置零
  // 然后我们使用第一行和第一列来记录其他行和列是否需要置零
  // 最后我们根据第一行和第一列的标记来置零
  const weightlength = matrix[0].length;
  const heightlength = matrix.length;
  let weight = new Array(weightlength).fill(false);
  let height = new Array(heightlength).fill(false);
  for (let i = 0; i < heightlength; i++) {
    for (let j = 0; j < weightlength; j++) {
      if (matrix[i][j] === 0) {
        weight[j] = true;
        height[i] = true;
      }
    }
  }
  for (let i = 0; i < heightlength; i++) {
    for (let j = 0; j < weightlength; j++) {
      if (weight[j] || height[i]) {
        matrix[i][j] = 0;
      }
    }
  }
  return matrix;
};
